mailpoet/assets/js/src/automation/listing/index.tsx

```diff
@@ -1,17 +1,90 @@
 import { Search, TableCard } from '@woocommerce/components/build';
+import { TabPanel } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import { useHistory, useLocation } from 'react-router-dom';
 import { getRow } from './get-row';
-import { Workflow } from './workflow';
+import { Workflow, WorkflowStatus } from './workflow';
 
 type Props = {
   workflows: Workflow[];
   loading: boolean;
 };
 
+type FilterTab = {
+  name: string;
+  title?: JSX.Element;
+  classname: string;
+};
+
 export function AutomationListing({ workflows, loading }: Props): JSX.Element {
-  const rows = workflows.map((workflow) => getRow(workflow));
+  const history = useHistory();
+  const location = useLocation();
+  const pageSearch = new URLSearchParams(location.search);
+
+  const updateUrlSearchString = (keysAndValues: Record<string, string>) => {
+    Object.keys(keysAndValues).forEach((key) => {
+      pageSearch.set(key, keysAndValues[key]);
+      if (['per_page', 'status'].includes(key)) {
+        pageSearch.set('paged', '1');
+      }
+    });
+    history.replace({ search: pageSearch.toString() });
+  };
+
+  const groupedWorkflows = {};
+  workflows.forEach((workflow) => {
+    if (!groupedWorkflows[workflow.status]) {
+      groupedWorkflows[workflow.status] = [];
+    }
+    groupedWorkflows[workflow.status].push(workflow);
+  });
```

Comment:
Here, I'd probably use `useMemo` to avoid recomputing it on every render.

There's also a declarative way to write this, although not sure if more readable:
```ts
const groupedWorkflows = workflows.reduce(
  (grouped, workflow) => ({
    ...grouped,
    [workflow.status]: [...(grouped[workflow.status] ?? []), workflow],
  }),
  {},
);
```

Reply:
Good call on `useMemo`!

This is mostly a personal failing on my part, but I've never liked using `reduce`. It was hard for me to understand when I first started coding and I've never warmed up to it.

In this case I'm curious, would the performance be the same using the spread syntax compared to `push`? I did a little bit of testing and I think spread option is slower, but it probably wouldn't make any kind of noticeable difference without there being tons and tons of workflows to filter.

I tested with 15k workflows and your code took ~700ms to sort compared to ~1ms for using `.push`. The much, much slower part was actually the query to retrieve all the workflows in the first place because there's no `workflow_id` index on the `workflow_versions` table, so I'm going to add one.

I'm happy to change to something more declarative but I'd like to avoid the performance hit if possible.